{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os,csv,re\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import math\n",
    "import SpaGCN as spg\n",
    "from scipy.sparse import issparse\n",
    "import random, torch\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import matplotlib.colors as clr\n",
    "import matplotlib.pyplot as plt\n",
    "import SpaGCN as spg\n",
    "import cv2\n",
    "import os\n",
    "import sys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.metrics.cluster import adjusted_rand_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# the location of R (used for the mclust clustering)\n",
    "# os.environ['R_HOME'] = 'D:\\Program Files\\R\\R-4.0.3'\n",
    "os.environ['R_HOME'] = '/home/miniconda3/envs/3.7/lib/R/'\n",
    "# os.environ['R_USER'] = 'D:\\ProgramData\\Anaconda3\\Lib\\site-packages\\rpy2'\n",
    "os.environ['R_USER'] = '/home/miniconda3/envs/3.7/lib/python3.7/site-packages/rpy2'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Read in gene expression and spatial location\n",
    "adata1=sc.read(\"../dataset/DLPFC/151673/sample_data.h5ad\")\n",
    "#Read in hitology image\n",
    "img=cv2.imread(\"../dataset/DLPFC/151673/histology.tif\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "section_id = '151673'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_dir = os.path.join('../dataset/DLPFC', section_id)\n",
    "adata = sc.read_visium(path=input_dir, count_file=section_id+'_filtered_feature_bc_matrix.h5')\n",
    "adata.var_names_make_unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs = adata.obs.rename(columns={'array_row': 'x_array'})\n",
    "adata.obs = adata.obs.rename(columns={'array_col': 'y_array'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs[\"x_pixel\"]=adata1.obs[\"x4\"]\n",
    "adata.obs[\"y_pixel\"]=adata1.obs[\"x5\"]\n",
    "x_array=adata.obs[\"x_array\"].tolist()\n",
    "y_array=adata.obs[\"y_array\"].tolist()\n",
    "x_pixel=adata.obs[\"x_pixel\"].tolist()\n",
    "y_pixel=adata.obs[\"y_pixel\"].tolist()\n",
    "\n",
    "#Test coordinates on the image\n",
    "img_new=img.copy()\n",
    "for i in range(len(x_pixel)):\n",
    "    x=x_pixel[i]\n",
    "    y=y_pixel[i]\n",
    "    img_new[int(x-20):int(x+20), int(y-20):int(y+20),:]=0\n",
    "\n",
    "cv2.imwrite('../dataset/DLPFC/151673/map.jpg', img_new)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#将adata存成raw文件(.h5ad)用于识别SVG\n",
    "adata.write(\"../dataset/DLPFC/151673/processed_data.h5ad\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Normalization\n",
    "sc.pp.highly_variable_genes(adata, flavor=\"seurat_v3\", n_top_genes=3000)  \n",
    "sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "sc.pp.log1p(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read the annotation\n",
    "Ann_df = pd.read_csv(os.path.join('../dataset/DLPFC', section_id, section_id+'_truth.txt'), sep='\\t', header=None, index_col=0)\n",
    "Ann_df.columns = ['Ground Truth']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Constructing the spatial network"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:From /home/tjzhang02/miniconda3/envs/3.7/lib/python3.7/site-packages/tensorflow_core/python/compat/v2_compat.py:68: disable_resource_variables (from tensorflow.python.ops.variable_scope) is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "non-resource variables are not supported in the long term\n"
     ]
    }
   ],
   "source": [
    "import utils\n",
    "import Train_STAGATE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "------Calculating spatial graph...\n",
      "The graph contains 21124 edges, 3639 cells.\n",
      "5.8049 neighbors per cell on average.\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAU0AAADcCAYAAAD0vfOWAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/NK7nSAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAtaUlEQVR4nO3de1xM+f8H8Nd0nXQZii7SdlEoIkpJKETrtqxbK5S+xPpmXVp+tIuwVNj6Zt2SdYnV6uv6tVi+acOyrUvxxa7kltroJkoX3ebz+8N3zteYSZ2RZvB+Ph7zeDSf8/l85n3Omd7zOed8zoyAMcZACCGkUdSUHQAhhLxLKGkSQggPlDQJIYQHSpqEEMIDJU1CCOGBkiYhhPBASZMQQnigpEkIITxQ0iSEEB4oaTbg9OnTEAgE2L9/v7JDaZT8/HyMHTsWRkZGEAgEiImJafYYJNvs9OnTCrdtzPb28vJCly5dFIiw+V28eBFaWlp48OCBskMhAGpqamBhYYFNmzbxbqsSSXPnzp0QCAQQCoXIzc2VWf4u/XMo27x583Dy5EmEhoZi9+7d+Pjjj+utKxAIIBAIEBUVJbNMsk8uX778NsP9YHz99deYMGECLC0tuTIvLy8IBALY2dnJbZOUlMTto3flQ7s+y5Yt49bl5YdQKGxUe7FYjNjYWDg5OUFPTw8mJiYYMmQIfvvtN5m6VVVVWLhwIdq2bQsdHR24ubkhKSlJqo6mpiZCQkKwatUqPH/+nNe6qETSlKiqqkJkZKSyw3in/fLLLxg5ciTmz5+PSZMmoVOnTg22Wbt2LSoqKposhn79+qGyshL9+vVrsj7fZVevXsWpU6fw+eefyywTCoW4c+cOLl68KLNsz549jU4q74rNmzdj9+7d3GPHjh2NardgwQLMnDkTjo6OiI6OxpdffonMzEx4enrKbLspU6YgOjoaEydOxLp166Curo6hQ4fi3LlzUvUCAwNRVFSEhIQEfivBVMCOHTsYAObk5MS0tbVZbm6u1HJPT0/WuXNnpcSWkpLCALB9+/a91dcpKytrkn4EAgELDg5uVF3JNgfAoqKipJZJ9smlS5eaJK7G4rO9m+N9UVNTw6qqqt6oj9mzZ7OPPvqIicViqXJJ/B07dmRz586VWlZZWckMDAzYmDFjmuX997aFhYUxAKywsJB325qaGqajo8PGjh0rVX7v3j0GgM2ePZsru3DhAgPA1q5dy5VVVlay9u3bM3d3d5m+hw8fzvr27csrHpUaaX711Veoq6trcLSZlZUFgUCAnTt3yiwTCARYtmwZ91xyWJCZmYlJkyZBJBKhTZs2WLJkCRhjyMnJwciRI2FgYABTU1O5h6oAUFdXh6+++gqmpqbQ1dXFJ598gpycHJl6Fy5cwMcffwyRSIQWLVrA09MT58+fl6ojienPP/+En58fWrVqhT59+rx2ne/du4dx48bB0NAQLVq0QK9evXDs2DFuueRwmjGGjRs3coc/DfHw8MCAAQOwZs0aVFZWNlg/IyMDY8eOhaGhIYRCIVxcXHDkyBGpOvWd09y4cSNsbGygo6MDV1dX/Prrr/Dy8oKXl5fM64jFYqxatQrt2rWDUCjEwIEDcefOHbkxpaWloXfv3tDR0YG1tTViY2Nl6hQUFGDq1KkwMTGBUChEt27dEB8fL1VH8r769ttvERMTg/bt20NbWxt//vknAGD9+vXo3LkzWrRogVatWsHFxaVRo5TDhw9jwIAB9e6PCRMmIDExEWKxmCv76aefUFFRgfHjx8ttk5ubi7/97W8wMTGBtrY2OnfujO3bt0vVqa6uxtKlS+Hs7AyRSARdXV307dsXKSkp9a53XFwct949e/bEpUuXGlw/PhhjKC0tBePx5Wo1NTWorKyEiYmJVLmxsTHU1NSgo6PDle3fvx/q6uqYPn06VyYUCjF16lSkpqbK/M8OGjQI586dQ3FxcaPjUamkaW1tDX9/f2zduhUPHz5s0r59fX0hFosRGRkJNzc3rFy5EjExMRg0aBDMzc2xevVq2NraYv78+Th79qxM+1WrVuHYsWNYuHAhZs+ejaSkJHh7e0slml9++QX9+vVDaWkpwsLCEB4ejqdPn2LAgAFyD7/GjRuHiooKhIeHIygoqN7Y8/Pz0bt3b5w8eRJ///vfufMwn3zyCQ4dOgTgxSHx7t27Abx4I0gOfxpj2bJlyM/Px+bNm19b748//kCvXr1w8+ZNLFq0CFFRUdDV1cWoUaO4OOqzefNmzJo1C+3atcOaNWvQt29fjBo1Cn/99Zfc+pGRkTh06BDmz5+P0NBQ/P7775g4caJMvSdPnmDo0KFwdnbGmjVr0K5dO8ycOVMqgVRWVsLLywu7d+/GxIkTsXbtWohEIkyZMgXr1q2T6XPHjh1Yv349pk+fjqioKBgaGmLr1q2YPXs2HBwcEBMTg+XLl8PJyQkXLlx47Xrn5uYiOzsbPXr0qLeOn58fHj16JPUhk5CQgIEDB8LY2Fimfn5+Pnr16oVTp05h1qxZWLduHWxtbTF16lSpC3+lpaX4/vvv4eXlhdWrV2PZsmUoLCyEj48Prl69KtNvQkIC1q5dixkzZmDlypXIysrC6NGjUVNTw9WpqqpCUVFRox7y2NjYQCQSQV9fH5MmTUJ+fv5rtx8A7rzkzp07sWfPHmRnZ+PatWuYMmUKWrVqJZUgr1y5gg4dOsDAwECqD1dXVwCQWW9nZ2cwxuSeG60Xr3HpW/LyoeDdu3eZhoaG1JD71cOw+/fvMwBsx44dMn0BYGFhYdxzyWHB9OnTubLa2lrWrl07JhAIWGRkJFf+5MkTpqOjwwICArgyyeGiubk5Ky0t5cr/+c9/MgBs3bp1jDHGxGIxs7OzYz4+PlKHYRUVFcza2poNGjRIJqYJEyY0avvMnTuXAWC//vorV/bs2TNmbW3NrKysWF1dndT68zk8l9Tt378/MzU1ZRUVFYwx+YfnAwcOZI6Ojuz58+dcmVgsZr1792Z2dnZcmWSbpaSkMMYYq6qqYkZGRqxnz56spqaGq7dz504GgHl6esq0tbe3lzosXrduHQPArl+/zpV5enrKnFqoqqpiTk5OzNjYmFVXVzPGGIuJiWEA2A8//MDVq66uZu7u7kxPT4/br5L3lYGBASsoKJDaViNHjlToVMCpU6cYAPbTTz/JLHv5fe3i4sKmTp3KGHvxPtTS0mLx8fFyT1dMnTqVmZmZsaKiIqn+PvvsMyYSibh9WFtbK3Nq4cmTJ8zExIT97W9/48ok621kZMSKi4u58n/9618ysUveF415vCwmJobNmjWL7dmzh+3fv5/NmTOHaWhoMDs7O1ZSUtLgdrx9+zbr0aOHVP82NjYsIyNDql7nzp3ZgAEDZNr/8ccfDACLjY2VKn/48CEDwFavXt1gDBIqNdIEXnwSTZ48GXFxcXj06FGT9Ttt2jTub3V1dbi4uIAxhqlTp3LlLVu2RMeOHXHv3j2Z9v7+/tDX1+eejx07FmZmZjh+/DiAF59gt2/fhp+fHx4/fsx92paXl2PgwIE4e/as1OEXALkXBuQ5fvw4XF1dpQ7h9fT0MH36dGRlZXGHj29i2bJlyMvLk3toCwDFxcX45ZdfMH78eDx79oxbv8ePH8PHxwe3b9+WO/MBAC5fvozHjx8jKCgIGhoaXPnEiRPRqlUruW0CAwOhpaXFPe/bty8AyOwbDQ0NzJgxg3uupaWFGTNmoKCgAGlpaQBebD9TU1NMmDCBq6epqYnZs2ejrKwMZ86ckepzzJgxaNOmjVRZy5Yt8ddff/E+XH38+DEA1LueEn5+fjh48CCqq6u5Q8xPP/1Uph5jDAcOHMCIESPAGJMa2fn4+KCkpATp6ekAXrzPJdtQLBajuLgYtbW1cHFx4eq8zNfXVypOedvcx8cHSUlJjXq8bM6cOVi/fj38/PwwZswYxMTEID4+Hrdv327UtB99fX107twZwcHBOHjwIDZt2oTa2lqMGjVKalRbWVkJbW1tmfaSC2qvnoKSrG99I2N5NBqu0vwWL16M3bt3IzIyUu7hkyI++ugjqecikQhCoRCtW7eWKZe80V/26rQQgUAAW1tbZGVlAQBu374NAAgICKg3hpKSEqk3pbW1daNif/DgAdzc3GTK7e3tueVvOiWrX79+6N+/P9asWSM3md+5cweMMSxZsgRLliyR20dBQQHMzc3lxg8Atra2UuUaGhqwsrKS29er+0uy3Z48eSJV3rZtW+jq6kqVdejQAcCLc3W9evXCgwcPYGdnBzU16THCy9vvZfL2y8KFC3Hq1Cm4urrC1tYWgwcPhp+fHzw8POTG/yrWwDm8zz77DPPnz8fPP/+MPXv2YPjw4VIf0hKFhYV4+vQp4uLiEBcXJ7evgoIC7u/4+HhERUUhIyND6jBb3jo2ZpubmZnBzMzstevSWH5+fvjyyy9x6tQpLFq0qN56tbW18Pb2hpeXF9avX8+Ve3t7o3Pnzli7di1Wr14N4MWhfFVVlUwfkmlFL5//BP63Xxpz/l9CJZOmjY0NJk2ahLi4OLkbs74VrKurq7dPdXX1RpUBDb/B5ZGMIteuXQsnJye5dfT09KSev7oDlS0sLAxeXl7YsmULWrZsKbVMsn7z58+Hj4+P3PavJsU30ZT7hi95+8Xe3h63bt3C0aNHceLECRw4cACbNm3C0qVLsXz58nr7MjIyAiCb7F9lZmYGLy8vREVF4fz58zhw4IDcepL9MGnSpHo/oLt27QoA+OGHHzBlyhSMGjUKCxYsgLGxMdTV1REREYG7d+/KtGvMNq+srERJSclr10XC1NS0wToWFhYNXoQ5e/Ysbty4gejoaKlyOzs72NvbS11oNTMzk3vEIzlqbdu2rVS5ZL+8Onh6HZVMmsCL0eYPP/zAfYK8TPIJ+PTpU6nyt3m3hWQkKcEYw507d7g3aPv27QEABgYG8Pb2btLXtrS0xK1bt2TKMzIyuOVNwdPTk7tosHTpUqllNjY2AF4c1vJdP0l8d+7cQf/+/bny2tpaZGVlcdtQEQ8fPkR5ebnUaDMzMxMAuFGspaUlrl27BrFYLDXa5Lv9dHV14evrC19fX1RXV2P06NFYtWoVQkND651PKZkne//+/Qb79/Pzw7Rp09CyZUsMHTpUbp02bdpAX18fdXV1De6H/fv3w8bGBgcPHpQaaISFhTUYS30SExMRGBjYqLoNfcAxxpCVlYXu3bu/tp7kYpG8QVFNTQ1qa2u5505OTkhJSUFpaanUxSDJBbtXBzSS/SI56mgMlTunKdG+fXtMmjQJW7ZsQV5entQyAwMDtG7dWuYqtyK3RDXWrl278OzZM+75/v378ejRIwwZMgTAi6tw7du3x7fffouysjKZ9oWFhQq/9tChQ3Hx4kWkpqZyZeXl5YiLi4OVlRUcHBwU7vtVknObrx76GRsbc6NQeeeaX7d+Li4uMDIywtatW6Xe4Hv27GlwBNaQ2tpabNmyhXteXV2NLVu2oE2bNnB2dgbwYvvl5eUhMTFRqt369euhp6cHT0/PBl/n1VM2WlpacHBwAGNM6rD3Vebm5rCwsGjUnVVjx45FWFgYNm3aJHU+92Xq6uoYM2YMDhw4gBs3bsgsf3k/SEaOLyevCxcuSL2P+FL0nKa898fmzZtRWFgoc9daRkYGsrOzueeS0y179+6Vqpeeno5bt25JJd2xY8eirq5O6v1bVVWFHTt2wM3NDRYWFlJ9pKWlQSAQwN3dvdHbQGVHmsCLW892796NW7duoXPnzlLLpk2bhsjISEybNg0uLi44e/YsN8J4GwwNDdGnTx8EBgYiPz8fMTExsLW15aYKqamp4fvvv8eQIUPQuXNnBAYGwtzcHLm5uUhJSYGBgQF++uknhV570aJF+PHHHzFkyBDMnj0bhoaGiI+Px/3793HgwAGZc3VvwtPTE56enjIXR4AX8yz79OkDR0dHBAUFwcbGBvn5+UhNTcVff/2F//znP3L71NLSwrJly/DFF19gwIABGD9+PLKysrBz5060b9+e1/mkV7Vt2xarV69GVlYWOnTogMTERFy9ehVxcXHQ1NQEAEyfPh1btmzBlClTkJaWBisrK+zfvx/nz59HTEyM3HOHrxo8eDBMTU3h4eEBExMT3Lx5Exs2bMCwYcMabD9y5EgcOnQIjLHXrqtIJJKaY1yfyMhIpKSkwM3NDUFBQXBwcEBxcTHS09Nx6tQp7nB3+PDhOHjwID799FMMGzYM9+/fR2xsLBwcHOR+sDeGouc0LS0t4evrC0dHRwiFQpw7dw579+6Fk5OT1IU84MWoz9PTk5uC5ezsjEGDBiE+Ph6lpaUYPHgwHj16hPXr10NHRwdz587l2rq5uWHcuHEIDQ1FQUEBbG1tER8fj6ysLGzbtk0mrqSkJHh4eHCnURql0dfZ36LX3X0SEBDAAMhM96ioqGBTp05lIpGI6evrs/Hjx7OCgoJ6pxy9eidCQEAA09XVlXm9V6c3SaZ8/Pjjjyw0NJQZGxszHR0dNmzYMPbgwQOZ9leuXGGjR49mRkZGTFtbm1laWrLx48ez5OTkBmN6nbt377KxY8eyli1bMqFQyFxdXdnRo0dl6kHBKUcvk6yzvH1y9+5d5u/vz0xNTZmmpiYzNzdnw4cPZ/v375dpL5lyJPHdd98xS0tLpq2tzVxdXdn58+eZs7Mz+/jjj2XavnoHjLxpZpJ9dfnyZebu7s6EQiGztLRkGzZskFmn/Px8FhgYyFq3bs20tLSYo6OjzJQ1yWu8fDeJxJYtW1i/fv24/dq+fXu2YMGCRk2XSU9Pl5ky9nL8r1Pf9sjPz2fBwcHMwsKCaWpqMlNTUzZw4EAWFxfH1RGLxSw8PJzb5t27d2dHjx5lAQEBzNLSslHr/er/k6KmTZvGHBwcmL6+PtPU1GS2trZs4cKFUtP4Xn7Nl6ehMfbi/33FihXMwcGB6ejoMJFIxIYPH86uXLki076yspLNnz+fmZqaMm1tbdazZ0924sQJmXpPnz5lWlpa7Pvvv+e1LoL/BklIsxOLxWjTpg1Gjx6NrVu3Kjuct2rgwIFo27Zto284IG9fTEwM1qxZg7t37/K6KKuy5zTJ++X58+cyFwZ27dqF4uJiubdRvm/Cw8ORmJhIXw2nImpqahAdHY3FixfznsVCI03SLE6fPo158+Zh3LhxMDIyQnp6OrZt2wZ7e3ukpaXVe+GDEFWj0heCyPvDysoKFhYW+O6771BcXAxDQ0P4+/sjMjKSEiZ5p9BIkxBCeKBzmoQQwgMlTUII4eGDO6cpFovx8OFD6Ovrv9GkakLI28EYw7Nnz9C2bdsmvXGjqXxwSfPhw4cyt1IRQlRPTk4O2rVrp+wwZHxwSVNyy1tOTo7MtzsTQpSvtLQUFhYWjbq9VRk+uKQpOSQ3MDCgpEmIClPV02eqd8KAEEJUGCVNQgjhgZImIYTwQEmTEEJ4oKRJCCE8fHBXzwkhgNWiY83+mlmRw5r9Nd8GGmkSQggPlDQJIYQHSpqEEMIDJU1CCOGBkiYhhPBASZMQQnigpEkIITxQ0iSEEB4oaRJCCA+UNAkhhAdKmoQQwgMlTUII4UHpSXPjxo2wsrKCUCiEm5sbLl68+Nr6MTEx6NixI3R0dGBhYYF58+bh+fPnzRQtIeRDp9SkmZiYiJCQEISFhSE9PR3dunWDj48PCgoK5NZPSEjAokWLEBYWhps3b2Lbtm1ITEzEV1991cyRE0I+VEpNmtHR0QgKCkJgYCAcHBwQGxuLFi1aYPv27XLr//bbb/Dw8ICfnx+srKwwePBgTJgwocHRKSGENBWlJc3q6mqkpaXB29v7f8GoqcHb2xupqaly2/Tu3RtpaWlckrx37x6OHz+OoUOH1vs6VVVVKC0tlXoQQoiilPYlxEVFRairq4OJiYlUuYmJCTIyMuS28fPzQ1FREfr06QPGGGpra/H555+/9vA8IiICy5cvb9LYCSEfLqVfCOLj9OnTCA8Px6ZNm5Ceno6DBw/i2LFj+Oabb+ptExoaipKSEu6Rk5PTjBETQt43Shtptm7dGurq6sjPz5cqz8/Ph6mpqdw2S5YsweTJkzFt2jQAgKOjI8rLyzF9+nR8/fXXUFOT/QzQ1taGtrZ2068AIeSDpLSRppaWFpydnZGcnMyVicViJCcnw93dXW6biooKmcSorq4OAGCMvb1gCSHkv5T6w2ohISEICAiAi4sLXF1dERMTg/LycgQGBgIA/P39YW5ujoiICADAiBEjEB0dje7du8PNzQ137tzBkiVLMGLECC55EkLI26TUpOnr64vCwkIsXboUeXl5cHJywokTJ7iLQ9nZ2VIjy8WLF0MgEGDx4sXIzc1FmzZtMGLECKxatUpZq0AI+cAI2Ad2XFtaWgqRSISSkhIYGBgoOxxClEKVf8JX1f9H36mr54QQomyUNAkhhAdKmoQQwgMlTUII4YGSJiGE8EBJkxBCeKCkSQghPFDSJIQQHihpEkIIDwonzadPn+L7779HaGgoiouLAQDp6enIzc1tsuAIIUTVKHTv+bVr1+Dt7Q2RSISsrCwEBQXB0NAQBw8eRHZ2Nnbt2tXUcRJCiEpQaKQZEhKCKVOm4Pbt2xAKhVz50KFDcfbs2SYLjhBCVI1CSfPSpUuYMWOGTLm5uTny8vLeOChCCFFVCiVNbW1tuT9QlpmZiTZt2rxxUIQQoqoUSpqffPIJVqxYgZqaGgCAQCBAdnY2Fi5ciDFjxjRpgIQQokoUSppRUVEoKyuDsbExKisr4enpCVtbW+jr69MXAhNC3msKXT0XiURISkrCuXPncO3aNZSVlaFHjx5Sv2FOCCHvozf6uYs+ffqgT58+TRULIYSoPIWS5nfffSe3XCAQQCgUwtbWFv369aMfOyOEvHcUSpr/+Mc/UFhYiIqKCrRq1QoA8OTJE7Ro0QJ6enooKCiAjY0NUlJSYGFh0aQBE0KIMil0ISg8PBw9e/bE7du38fjxYzx+/BiZmZlwc3PDunXrkJ2dDVNTU8ybN6+p4yWEEKVSKGkuXrwY//jHP9C+fXuuzNbWFt9++y1CQ0PRrl07rFmzBufPn2+wr40bN8LKygpCoRBubm64ePHia+s/ffoUwcHBMDMzg7a2Njp06IDjx48rshqEEMKbQofnjx49Qm1trUx5bW0td0dQ27Zt8ezZs9f2k5iYiJCQEMTGxsLNzQ0xMTHw8fHBrVu3YGxsLFO/uroagwYNgrGxMfbv3w9zc3M8ePAALVu2VGQ1CCGEN4VGmv3798eMGTNw5coVruzKlSuYOXMmBgwYAAC4fv06rK2tX9tPdHQ0goKCEBgYCAcHB8TGxqJFixbYvn273Prbt29HcXExDh8+DA8PD1hZWcHT0xPdunVTZDUIIYQ3hZLmtm3bYGhoCGdnZ2hra0NbWxsuLi4wNDTEtm3bAAB6enqIioqqt4/q6mqkpaVJze1UU1ODt7c3UlNT5bY5cuQI3N3dERwcDBMTE3Tp0gXh4eGoq6tTZDUIIYQ3hQ7PTU1NkZSUhIyMDGRmZgIAOnbsiI4dO3J1+vfv/9o+ioqKUFdXBxMTE6lyExMTZGRkyG1z7949/PLLL5g4cSKOHz+OO3fu4O9//ztqamoQFhYmt01VVRWqqqq45/LumSeEkMZ6o8ntnTp1QqdOnZoqlgaJxWIYGxsjLi4O6urqcHZ2Rm5uLtauXVtv0oyIiMDy5cubLUZCyPtN4aT5119/4ciRI8jOzkZ1dbXUsujo6Abbt27dGurq6sjPz5cqz8/Ph6mpqdw2ZmZm0NTUlJo0b29vj7y8PFRXV0NLS0umTWhoKEJCQrjnpaWlNHeUEKIwhZJmcnIyPvnkE9jY2CAjIwNdunRBVlYWGGPo0aNHo/rQ0tKCs7MzkpOTMWrUKAAvRpLJycmYNWuW3DYeHh5ISEiAWCyGmtqL07GZmZkwMzOTmzABcOdcCSGkKSh0ISg0NBTz58/H9evXIRQKceDAAeTk5MDT0xPjxo1rdD8hISHYunUr4uPjcfPmTcycORPl5eUIDAwEAPj7+yM0NJSrP3PmTBQXF2POnDnIzMzEsWPHEB4ejuDgYEVWgxBCeFNopHnz5k38+OOPLzrQ0EBlZSX09PSwYsUKjBw5EjNnzmxUP76+vigsLMTSpUuRl5cHJycnnDhxgrs4lJ2dzY0oAcDCwgInT57EvHnz0LVrV5ibm2POnDlYuHChIqtBCCG8KZQ0dXV1ufOYZmZmuHv3Ljp37gzgxVVxPmbNmlXv4fjp06dlytzd3fH777/zC5gQQpqIQkmzV69eOHfuHOzt7TF06FB8+eWXuH79Og4ePIhevXo1dYyEEKIyFEqa0dHRKCsrAwAsX74cZWVlSExMhJ2dXaOunBNCyLtKoaRpY2PD/a2rq4vY2NgmC4gQQlSZQlfPbWxs8PjxY5nyp0+fSiVUQgh53yiUNLOysuTe711VVYXc3Nw3DooQQlQVr8PzI0eOcH+fPHkSIpGIe15XV4fk5GRYWVk1WXCEEKJqeCVNyZ07AoEAAQEBUss0NTVhZWX12m82IoSQdx2vpCkWiwEA1tbWuHTpElq3bv1WgiKEEFWl0NXz+/fvN3UchBDyTlD4W46Sk5ORnJyMgoICbgQqUd83rxNCyLtOoaS5fPlyrFixAi4uLjAzM4NAIGjquAghRCUplDRjY2Oxc+dOTJ48uanjIYQQlabQPM3q6mr07t27qWMhhBCVp1DSnDZtGhISEpo6FkIIUXkKHZ4/f/4ccXFxOHXqFLp27QpNTU2p5fSlHYSQ95VCSfPatWtwcnICANy4cUNqGV0UIoS8zxRKmikpKU0dByGEvBMUOqcpcefOHZw8eRKVlZUAAMZYkwRFCCGqSqGk+fjxYwwcOBAdOnTA0KFD8ejRIwDA1KlT8eWXXzZpgIQQokoUSprz5s2DpqYmsrOz0aJFC67c19cXJ06caLLgCCFE1Sh0TvPf//43Tp48iXbt2kmV29nZ4cGDB00SGCGEqCKFRprl5eVSI0yJ4uJiaGtr8+5v48aNsLKyglAohJubGy5evNiodnv37oVAIOC+so4QQt42hZJm3759sWvXLu65QCCAWCzGmjVr0L9/f159JSYmIiQkBGFhYUhPT0e3bt3g4+ODgoKC17bLysrC/Pnz0bdvX0VWgRBCFKJQ0lyzZg3i4uIwZMgQVFdX4//+7//QpUsXnD17FqtXr+bVV3R0NIKCghAYGAgHBwfExsaiRYsWr/2mpLq6OkycOBHLly+n3yQihDQrhZJmly5dkJmZiT59+mDkyJEoLy/H6NGjceXKFbRv377R/VRXVyMtLQ3e3t7/C0hNDd7e3khNTa233YoVK2BsbIypU6cqEj4hhChM4e/TFIlE+Prrr9/oxYuKilBXVwcTExOpchMTE2RkZMhtc+7cOWzbtg1Xr15t1GtUVVWhqqqKe15aWqpwvIQQotBIc8eOHdi3b59M+b59+xAfH//GQdXn2bNnmDx5MrZu3dron9qIiIiASCTiHhYWFm8tPkLI+0+hpBkRESE3aRkbGyM8PLzR/bRu3Rrq6urIz8+XKs/Pz4epqalM/bt37yIrKwsjRoyAhoYGNDQ0sGvXLhw5cgQaGhq4e/euTJvQ0FCUlJRwj5ycnEbHRwghr1Lo8Dw7OxvW1tYy5ZaWlsjOzm50P1paWnB2dkZycjI3bUgsFiM5ORmzZs2Sqd+pUydcv35dqmzx4sV49uwZ1q1bJ3cUqa2trdA0KEIIkUehpGlsbIxr167J/Mb5f/7zHxgZGfHqKyQkBAEBAXBxcYGrqytiYmJQXl6OwMBAAIC/vz/Mzc0REREBoVCILl26SLVv2bIlAMiUE0LI26BQ0pwwYQJmz54NfX199OvXDwBw5swZzJkzB5999hmvvnx9fVFYWIilS5ciLy8PTk5OOHHiBHdxKDs7G2pqb/S9IoQQ0mQETIGvJqqursbkyZOxb98+aGi8yLtisRj+/v6IjY2FlpZWkwfaVEpLSyESiVBSUgIDAwNlh0OIUlgtOtbsr5kVOaxR9VT9f5T3SJMxhry8POzcuRMrV67E1atXoaOjA0dHR1haWr6NGAkhRGUolDRtbW3xxx9/wM7ODnZ2dm8jLkIIUUm8TxaqqanBzs4Ojx8/fhvxEEKISlPoCktkZCQWLFgg8/tAhBDyvlPo6rm/vz8qKirQrVs3aGlpQUdHR2p5cXFxkwRHCCGqRqGkGRMT08RhEELIu0GhpBkQENDUcRBCyDtB4Vnjd+/exeLFizFhwgTuC4N//vln/PHHH00WHCGEqBqFkuaZM2fg6OiICxcu4ODBgygrKwPw4jbKsLCwJg2QEEJUiUJJc9GiRVi5ciWSkpKk7v4ZMGAAfv/99yYLjhBCVI1CSfP69ev49NNPZcqNjY1RVFT0xkERQoiqUihptmzZEo8ePZIpv3LlCszNzd84KEIIUVUKJc3PPvsMCxcuRF5eHvdLlOfPn8f8+fPh7+/f1DESQojKUChphoeHw97eHh999BHKysrg4OCAfv36oXfv3li8eHFTx0gIISqD1zxNsViMtWvX4siRI9zXw40ZMwZlZWXo3r07fXkHIeS9xytprlq1CsuWLYO3tzd0dHSQkJAAxthrf6OcEELeJ7wOz3ft2oVNmzbh5MmTOHz4MH766Sfs2bMHYrH4bcVHCCEqhVfSzM7OxtChQ7nn3t7eEAgEePjwYZMHRgghqohX0qytrYVQKJQq09TURE1NTZMGRQghqorXOU3GGKZMmSL1k7jPnz/H559/Dl1dXa7s4MGDTRchIYSoEF5JU963G02aNKnJgiGEEFXHK2nu2LHjbcVBCCHvBJX4QfGNGzfCysoKQqEQbm5uuHjxYr11t27dir59+6JVq1Zo1aoVvL29X1ufEEKaktKTZmJiIkJCQhAWFob09HR069YNPj4+3Hd0vur06dOYMGECUlJSkJqaCgsLCwwePBi5ubnNHDkh5EMkYIwxZQbg5uaGnj17YsOGDQBe3HVkYWGBL774AosWLWqwfV1dHVq1aoUNGzY06r53Vf8hekKag9WiY83+mlmRwxpVT9X/R5U60qyurkZaWhq8vb25MjU1NXh7eyM1NbVRfVRUVKCmpgaGhoZyl1dVVaG0tFTqQQghilJq0iwqKkJdXR1MTEykyk1MTJCXl9eoPhYuXIi2bdtKJd6XRUREQCQScQ8LC4s3jpsQ8uFS+jnNNxEZGYm9e/fi0KFDMpPuJUJDQ1FSUsI9cnJymjlKQsj7RKFfo2wqrVu3hrq6OvLz86XK8/PzYWpq+tq23377LSIjI3Hq1Cl07dq13nra2tpSk/EJIeRNKHWkqaWlBWdnZyQnJ3NlYrEYycnJcHd3r7fdmjVr8M033+DEiRNwcXFpjlAJIQSAkkeaABASEoKAgAC4uLjA1dUVMTExKC8vR2BgIADA398f5ubmiIiIAACsXr0aS5cuRUJCAqysrLhzn3p6etDT01PaehBCPgxKT5q+vr4oLCzE0qVLkZeXBycnJ5w4cYK7OJSdnQ01tf8NiDdv3ozq6mqMHTtWqp+wsDAsW7asOUMnhHyAlD5Ps7mp+hwwQpoDzdNU3Dt99ZwQQpobJU1CCOGBkiYhhPBASZMQQnigpEkIITxQ0iSEEB4oaRJCCA+UNAkhhAdKmoQQwgMlTUII4YGSJiGE8EBJkxBCeKCkSQghPFDSJIQQHihpEkIID5Q0CSGEB0qahBDCAyVNQgjhgZImIYTwoPQfViOkqany79+Qdx+NNAkhhAeVSJobN26ElZUVhEIh3NzccPHixdfW37dvHzp16gShUAhHR0ccP368mSIlhHzolH54npiYiJCQEMTGxsLNzQ0xMTHw8fHBrVu3YGxsLFP/t99+w4QJExAREYHhw4cjISEBo0aNQnp6Orp06aKENSB0ONww2kbvD6WPNKOjoxEUFITAwEA4ODggNjYWLVq0wPbt2+XWX7duHT7++GMsWLAA9vb2+Oabb9CjRw9s2LChmSMnhHyIlDrSrK6uRlpaGkJDQ7kyNTU1eHt7IzU1VW6b1NRUhISESJX5+Pjg8OHDcutXVVWhqqqKe15SUgLgxQ/Sk6Yhrqpo9td83f5TtXgA1YtJ1eKRV48x9jbDUZhSk2ZRURHq6upgYmIiVW5iYoKMjAy5bfLy8uTWz8vLk1s/IiICy5cvlym3sLBQMGqiCkQxyo5AmqrFA6heTHzjefbsGUQi0VuJ5U0o/Zzm2xYaGio1MhWLxSguLoaRkREEAoESI2tepaWlsLCwQE5ODgwMDJQdjkqibdSw5thGjDE8e/YMbdu2fSv9vymlJs3WrVtDXV0d+fn5UuX5+fkwNTWV28bU1JRXfW1tbWhra0uVtWzZUvGg33EGBgaUEBpA26hhb3sbqeIIU0KpF4K0tLTg7OyM5ORkrkwsFiM5ORnu7u5y27i7u0vVB4CkpKR66xNCSFNS+uF5SEgIAgIC4OLiAldXV8TExKC8vByBgYEAAH9/f5ibmyMiIgIAMGfOHHh6eiIqKgrDhg3D3r17cfnyZcTFxSlzNQghHwilJ01fX18UFhZi6dKlyMvLg5OTE06cOMFd7MnOzoaa2v8GxL1790ZCQgIWL16Mr776CnZ2djh8+DDN0WyAtrY2wsLCZE5VkP+hbdQw2kaAgKnqdX1CCFFBSp/cTggh7xJKmoQQwgMlTUII4YGSJiGE8EBJ8z0WERGBnj17Ql9fH8bGxhg1ahRu3bql7LBUWmRkJAQCAebOnavsUFRKbm4uJk2aBCMjI+jo6MDR0RGXL19WdlhKQUnzPXbmzBkEBwfj999/R1JSEmpqajB48GCUl5crOzSVdOnSJWzZsgVdu3ZVdigq5cmTJ/Dw8ICmpiZ+/vln/Pnnn4iKikKrVq2UHZpS0JSjD0hhYSGMjY1x5swZ9OvXT9nhqJSysjL06NEDmzZtwsqVK+Hk5ISYmBhlh6USFi1ahPPnz+PXX39VdigqgUaaHxDJ1+IZGhoqORLVExwcjGHDhsHb21vZoaicI0eOwMXFBePGjYOxsTG6d++OrVu3KjsspaGk+YEQi8WYO3cuPDw86O6pV+zduxfp6encrbpE2r1797B582bY2dnh5MmTmDlzJmbPno34+Hhlh6YUSr+NkjSP4OBg3LhxA+fOnVN2KColJycHc+bMQVJSEoRCobLDUUlisRguLi4IDw8HAHTv3h03btxAbGwsAgIClBxd86OR5gdg1qxZOHr0KFJSUtCuXTtlh6NS0tLSUFBQgB49ekBDQwMaGho4c+YMvvvuO2hoaKCurk7ZISqdmZkZHBwcpMrs7e2RnZ2tpIiUi0aa7zHGGL744gscOnQIp0+fhrW1tbJDUjkDBw7E9evXpcoCAwPRqVMnLFy4EOrq6kqKTHV4eHjITFXLzMyEpaWlkiJSLkqa77Hg4GAkJCTgX//6F/T19bmfBBGJRNDR0VFydKpBX19f5hyvrq4ujIyM6Nzvf82bNw+9e/dGeHg4xo8fj4sXLyIuLu6D/TpGmnL0Hqvv5zx27NiBKVOmNG8w7xAvLy+acvSKo0ePIjQ0FLdv34a1tTVCQkIQFBSk7LCUgpImIYTwQBeCCCGEB0qahBDCAyVNQgjhgZImIYTwQEmTEEJ4oKRJCCE8UNIkhBAeKGkSQggPlDQJIYQHSpqEEMIDJU1CCOGBkiYhhPDw/0YO/AEfhGvOAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 300x200 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "utils.Cal_Spatial_Net(adata, rad_cutoff=150)\n",
    "utils.Stats_Spatial_Net(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Size of Input:  (3639, 3000)\n",
      "WARNING:tensorflow:From /home/tjzhang02/miniconda3/envs/3.7/lib/python3.7/site-packages/tensorflow_core/python/ops/math_grad.py:1375: where (from tensorflow.python.ops.array_ops) is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "Use tf.where in 2.0, which has the same broadcast rule as np.where\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-07-14 15:58:05.445009: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 AVX512F FMA\n",
      "2024-07-14 15:58:05.453408: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 2900000000 Hz\n",
      "2024-07-14 15:58:05.456019: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x5641f443c690 initialized for platform Host (this does not guarantee that XLA will be used). Devices:\n",
      "2024-07-14 15:58:05.456056: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version\n",
      "2024-07-14 15:58:05.457271: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcuda.so.1\n",
      "2024-07-14 15:58:05.479882: E tensorflow/stream_executor/cuda/cuda_driver.cc:318] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected\n",
      "2024-07-14 15:58:05.480205: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:169] retrieving CUDA diagnostic information for host: localhost.localdomain\n",
      "2024-07-14 15:58:05.480232: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:176] hostname: localhost.localdomain\n",
      "2024-07-14 15:58:05.480417: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:200] libcuda reported version is: 460.106.0\n",
      "2024-07-14 15:58:05.480538: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:204] kernel reported version is: 460.106.0\n",
      "2024-07-14 15:58:05.480564: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:310] kernel version seems to match DSO: 460.106.0\n",
      "100%|██████████| 500/500 [01:33<00:00,  5.36it/s]\n"
     ]
    }
   ],
   "source": [
    "adata = Train_STAGATE.train_STAGATE(adata, alpha=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3639, 33538)\n",
      "AnnData object with n_obs × n_vars = 3639 × 33538\n",
      "    obs: 'in_tissue', 'x_array', 'y_array', 'x_pixel', 'y_pixel', 'Ground Truth'\n",
      "    var: 'gene_ids', 'feature_types', 'genome', 'highly_variable', 'highly_variable_rank', 'means', 'variances', 'variances_norm'\n",
      "    uns: 'spatial', 'hvg', 'log1p', 'Ground Truth_colors', 'Spatial_Net'\n",
      "    obsm: 'spatial', 'STAGATE'\n"
     ]
    }
   ],
   "source": [
    "print(adata.shape)\n",
    "print(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]:                    __           __ \n",
      "   ____ ___  _____/ /_  _______/ /_\n",
      "  / __ `__ \\/ ___/ / / / / ___/ __/\n",
      " / / / / / / /__/ / /_/ (__  ) /_  \n",
      "/_/ /_/ /_/\\___/_/\\__,_/____/\\__/   version 6.0.0\n",
      "Type 'citation(\"mclust\")' for citing this R package in publications.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fitting ...\n",
      "  |======================================================================| 100%\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "在单细胞RNA测序数据上执行降维、计算邻居关系并进行聚类。最终adata对象将包含降维后的数据、UMAP坐标和聚类结果\n",
    "'''\n",
    "sc.pp.neighbors(adata, use_rep='STAGATE')\n",
    "sc.tl.umap(adata)\n",
    "adata = utils.mclust_R(adata, used_obsm='STAGATE', num_cluster=7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Adjusted rand index = 0.60\n"
     ]
    }
   ],
   "source": [
    "obs_df = adata.obs.dropna()\n",
    "ARI = adjusted_rand_score(obs_df['mclust'], obs_df['Ground Truth'])\n",
    "print('Adjusted rand index = %.2f' %ARI)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "mclust_array = adata.obs[\"mclust\"].to_numpy(dtype=np.int64)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Indentify SVGs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculateing adj matrix using histology image...\n",
      "Var of c0,c1,c2 =  33.30687202862215 174.55510595352243 46.84205750749746\n",
      "Var of x,y,z =  5606737.526317932 4468793.817921193 5606737.526317932\n"
     ]
    }
   ],
   "source": [
    "#Calculate adjacent matrix\n",
    "s=1\n",
    "b=49\n",
    "adj=spg.calculate_adj_matrix(x=x_pixel,y=y_pixel, x_pixel=x_pixel, y_pixel=y_pixel, image=img, beta=b, alpha=s, histology=True)\n",
    "np.savetxt('../dataset/DLPFC/151673/adj.csv', adj, delimiter=',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: adata.X seems to be already log-transformed.\n"
     ]
    }
   ],
   "source": [
    "#Expression data preprocessing\n",
    "adj=np.loadtxt('../dataset/DLPFC/151673/adj.csv', delimiter=',')\n",
    "adata.var_names_make_unique()\n",
    "spg.prefilter_genes(adata,min_cells=3) # avoiding all genes are zeros\n",
    "spg.prefilter_specialgenes(adata)\n",
    "#Normalize and take log for UMI\n",
    "sc.pp.normalize_per_cell(adata)\n",
    "sc.pp.log1p(adata)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**5.2 Set hyper-parameters\n",
    "p: Percentage of total expression contributed by neighborhoods.\n",
    "l: Parameter to control p.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Run 1: l [0.01, 1000], p [0.0, 153.8820492650696]\n",
      "Run 2: l [0.01, 500.005], p [0.0, 28.015447343094223]\n",
      "Run 3: l [0.01, 250.0075], p [0.0, 4.240330523308446]\n",
      "Run 4: l [0.01, 125.00874999999999], p [0.0, 0.5157276735032843]\n",
      "Run 5: l [62.509375, 125.00874999999999], p [0.028496868560644373, 0.5157276735032843]\n",
      "Run 6: l [93.7590625, 125.00874999999999], p [0.18753135107474428, 0.5157276735032843]\n",
      "Run 7: l [109.38390625, 125.00874999999999], p [0.32801349789332424, 0.5157276735032843]\n",
      "Run 8: l [117.196328125, 125.00874999999999], p [0.4156469508032292, 0.5157276735032843]\n",
      "Run 9: l [121.1025390625, 125.00874999999999], p [0.4640926787304587, 0.5157276735032843]\n",
      "Run 10: l [123.05564453125, 125.00874999999999], p [0.48950676051756026, 0.5157276735032843]\n",
      "recommended l =  124.032197265625\n"
     ]
    }
   ],
   "source": [
    "p=0.5 \n",
    "#Find the l value given p\n",
    "l=spg.search_l(p, adj, start=0.01, end=1000, tol=0.01, max_run=100)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "n_clusters: Number of spatial domains wanted.\n",
    "res: Resolution in the initial Louvain's Clustering methods. If the number of clusters is known, we can use the spg.search_res() fnction to search for suitable resolution(optional)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start at res =  0.7 step =  0.1\n",
      "Initializing cluster centers with louvain, resolution =  0.7\n",
      "Epoch  0\n",
      "Epoch  10\n",
      "Res =  0.7 Num of clusters =  6\n",
      "Initializing cluster centers with louvain, resolution =  0.7999999999999999\n",
      "Epoch  0\n",
      "Epoch  10\n",
      "Res =  0.7999999999999999 Num of clusters =  7\n",
      "recommended res =  0.7999999999999999\n"
     ]
    }
   ],
   "source": [
    "#For this data, we set the number of clusters=7 since this tissue has 7 layers\n",
    "n_clusters=7\n",
    "#Set seed\n",
    "r_seed=t_seed=n_seed=100\n",
    "#Search for suitable resolution\n",
    "res=spg.search_res(adata, adj, l, n_clusters, start=0.7, step=0.1, tol=5e-3, lr=0.05, max_epochs=20, r_seed=r_seed, t_seed=t_seed, n_seed=n_seed)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Initializing cluster centers with louvain, resolution =  0.7999999999999999\n",
      "Epoch  0\n",
      "Epoch  10\n",
      "Epoch  20\n",
      "Epoch  30\n",
      "Epoch  40\n",
      "Epoch  50\n",
      "Epoch  60\n",
      "Epoch  70\n",
      "Epoch  80\n",
      "Epoch  90\n",
      "Epoch  100\n",
      "Epoch  110\n",
      "Epoch  120\n",
      "Epoch  130\n",
      "Epoch  140\n",
      "Epoch  150\n",
      "Epoch  160\n",
      "Epoch  170\n",
      "Epoch  180\n",
      "Epoch  190\n"
     ]
    }
   ],
   "source": [
    "clf=spg.SpaGCN()\n",
    "clf.set_l(l)\n",
    "#Set seed\n",
    "random.seed(r_seed)\n",
    "torch.manual_seed(t_seed)\n",
    "np.random.seed(n_seed)\n",
    "#Run\n",
    "clf.train(adata,adj,init_spa=True,init=\"louvain\",res=res, tol=5e-3, lr=0.05, max_epochs=200)\n",
    "adata.obs[\"mclust\"]= mclust_array\n",
    "adata.obs[\"mclust\"]=adata.obs[\"mclust\"].astype('category')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Identify SVGs**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 3639 × 33538\n",
       "    obs: 'in_tissue', 'x_array', 'y_array', 'x_pixel', 'y_pixel'\n",
       "    var: 'gene_ids', 'feature_types', 'genome'\n",
       "    uns: 'spatial'\n",
       "    obsm: 'spatial'"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Read in raw data\n",
    "raw=sc.read(\"../dataset/DLPFC/151673/processed_data.h5ad\")\n",
    "raw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "raw.var_names_make_unique()\n",
    "raw.obs[\"mclust\"]=adata.obs[\"mclust\"].astype('category')\n",
    "#Convert sparse matrix to non-sparse\n",
    "raw.X=(raw.X.A if issparse(raw.X) else raw.X)\n",
    "sc.pp.log1p(raw)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rank_genes_groups(input_adata, target_cluster,nbr_list, label_col, adj_nbr=True, log=False):\n",
    "    if adj_nbr:\n",
    "        nbr_list=nbr_list+[target_cluster]\n",
    "        adata=input_adata[input_adata.obs[label_col].isin(nbr_list)]\n",
    "    else:\n",
    "        adata=input_adata.copy()\n",
    "    adata.var_names_make_unique()\n",
    "    adata.obs[\"target\"]=((adata.obs[label_col]==target_cluster)*1).astype('category')\n",
    "    # print(f'adata_obs_target:\\n{adata.obs[\"target\"]}')\n",
    "    sc.tl.rank_genes_groups(adata, groupby=\"target\",reference=\"rest\", n_genes=adata.shape[1],method='wilcoxon')\n",
    "    pvals_adj=[i[0] for i in adata.uns['rank_genes_groups'][\"pvals_adj\"]]\n",
    "    genes=[i[1] for i in adata.uns['rank_genes_groups'][\"names\"]]\n",
    "    # print('pvals_adj', '\\n', pvals_adj)\n",
    "    # print('genes', '\\n', genes)\n",
    "    if issparse(adata.X):\n",
    "        obs_tidy=pd.DataFrame(adata.X.A)\n",
    "    else:\n",
    "        obs_tidy=pd.DataFrame(adata.X)\n",
    "    obs_tidy.index=adata.obs[\"target\"].tolist()\n",
    "    obs_tidy.columns=adata.var.index.tolist()\n",
    "    obs_tidy=obs_tidy.loc[:,genes]\n",
    "    # print('obs_tidy:')\n",
    "    # print(obs_tidy.head())\n",
    "    # 1. compute mean value\n",
    "    mean_obs = obs_tidy.groupby(level=0).mean()\n",
    "    # 邻域内(0)某基因表达量的均值  目标内(1)某基因表达量的均值\n",
    "    # 2. compute fraction of cells having value >0\n",
    "    obs_bool = obs_tidy.astype(bool)\n",
    "    # print('obs_bool:')\n",
    "    # print(obs_bool.head())\n",
    "    # print('index layer num:', obs_bool.index.nlevels)\n",
    "    fraction_obs = obs_bool.groupby(level=0).sum() / obs_bool.groupby(level=0).count()\n",
    "    # 在邻域内表达某基因的细胞个数sum[0]/临域内的细胞总数count[0]  \n",
    "    # 在目标域内表达某基因的细胞个数sum[1]/目标域内的细胞总数count[1]  \n",
    "    # print('obs_bool:\\n', obs_bool)\n",
    "    # print('fraction_obs:sum & count\\n', obs_bool.groupby(level=0).sum(), obs_bool.groupby(level=0).count())\n",
    "    # print('fraction_obs:\\n', fraction_obs)\n",
    "    \n",
    "    # print('mean_obs.loc', mean_obs.loc[1], '\\n', mean_obs.loc[0])\n",
    "    # compute fold change.\n",
    "    if log: #The adata already logged\n",
    "        fold_change=np.exp((mean_obs.loc[1] - mean_obs.loc[0]).values)\n",
    "    else:\n",
    "        fold_change = (mean_obs.loc[1] / (mean_obs.loc[0]+ 1e-9)).values\n",
    "    df = {'genes': genes, 'in_group_fraction': fraction_obs.loc[1].tolist(), \"out_group_fraction\":fraction_obs.loc[0].tolist(),\"in_out_group_ratio\":(fraction_obs.loc[1]/fraction_obs.loc[0]).tolist(),\"in_group_mean_exp\": mean_obs.loc[1].tolist(), \"out_group_mean_exp\": mean_obs.loc[0].tolist(),\"fold_change\":fold_change.tolist(), \"pvals_adj\":pvals_adj}\n",
    "    df = pd.DataFrame(data=df)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_neighbor_clusters(target_cluster,cell_id, x, y, pred,radius, ratio=1/2):\n",
    "    cluster_num = dict()\n",
    "    for i in pred:\n",
    "        cluster_num[i] = cluster_num.get(i, 0) + 1\n",
    "    df = {'cell_id': cell_id, 'x': x, \"y\":y, \"pred\":pred}\n",
    "    df = pd.DataFrame(data=df)\n",
    "    df.index=df['cell_id']\n",
    "    target_df=df[df[\"pred\"]==target_cluster]\n",
    "    nbr_num={}\n",
    "    row_index=0\n",
    "    num_nbr=[]\n",
    "    for index, row in target_df.iterrows():\n",
    "        x=row[\"x\"]\n",
    "        y=row[\"y\"]\n",
    "        tmp_nbr=df[((df[\"x\"]-x)**2+(df[\"y\"]-y)**2)<=(radius**2)]\n",
    "        #tmp_nbr=df[(df[\"x\"]<x+radius) & (df[\"x\"]>x-radius) & (df[\"y\"]<y+radius) & (df[\"y\"]>y-radius)]\n",
    "        num_nbr.append(tmp_nbr.shape[0])\n",
    "        for p in tmp_nbr[\"pred\"]:\n",
    "            nbr_num[p]=nbr_num.get(p,0)+1\n",
    "    del nbr_num[target_cluster]\n",
    "    nbr_num_back=nbr_num.copy() #Backup\n",
    "    nbr_num=[(k, v)  for k, v in nbr_num.items() if v>(ratio*cluster_num[k])]\n",
    "    nbr_num.sort(key=lambda x: -x[1])\n",
    "    print(\"radius=\", radius, \"average number of neighbors for each spot is\", np.mean(num_nbr))\n",
    "    print(\" Cluster\",target_cluster, \"has neighbors:\")\n",
    "    for t in nbr_num:\n",
    "        print(\"Dmain \", t[0], \": \",t[1])\n",
    "    ret=[t[0] for t in nbr_num]\n",
    "    if len(ret)==0:\n",
    "        nbr_num_back=[(k, v)  for k, v in nbr_num_back.items()]\n",
    "        nbr_num_back.sort(key=lambda x: -x[1])\n",
    "        ret=[nbr_num_back[0][0]]\n",
    "        print(\"No neighbor domain found, only return one potential neighbor domain:\",ret)\n",
    "        print(\"Try bigger radius or smaller ratio.\")\n",
    "    return ret"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getXX(target):\n",
    "   \n",
    "    #Search radius such that each spot in the target domain has approximately 10 neighbors on average\n",
    "    adj_2d=spg.calculate_adj_matrix(x=x_array, y=y_array, histology=False)\n",
    "    start, end= np.quantile(adj_2d[adj_2d!=0],q=0.001), np.quantile(adj_2d[adj_2d!=0],q=0.1)\n",
    "    r=spg.search_radius(target_cluster=target, cell_id=adata.obs.index.tolist(), x=x_array, y=y_array, pred=adata.obs[\"mclust\"].tolist(), start=start, end=end, num_min=10, num_max=14,  max_run=100)\n",
    "    #Detect neighboring domains\n",
    "    nbr_domians=find_neighbor_clusters(target_cluster=target,\n",
    "                                    cell_id=raw.obs.index.tolist(), \n",
    "                                    x=raw.obs[\"x_array\"].tolist(), \n",
    "                                    y=raw.obs[\"y_array\"].tolist(), \n",
    "                                    pred=raw.obs[\"mclust\"].tolist(),\n",
    "                                    radius=r,\n",
    "                                    ratio=1/2)\n",
    "    return nbr_domians"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculateing adj matrix using xy only...\n",
      "Calculateing adj matrix using xy only...\n",
      "Calculateing adj matrix using xy only...\n",
      "Run 1: radius [1.4142135381698608, 16.970561981201172], num_nbr [1.0, 354.74418604651163]\n",
      "Calculateing adj matrix using xy only...\n",
      "Run 2: radius [1.4142135381698608, 9.192387759685516], num_nbr [1.0, 122.48172757475083]\n",
      "Calculateing adj matrix using xy only...\n",
      "Run 3: radius [1.4142135381698608, 5.303300648927689], num_nbr [1.0, 42.23421926910299]\n",
      "Calculateing adj matrix using xy only...\n",
      "Run 4: radius [1.4142135381698608, 3.3587570935487747], num_nbr [1.0, 20.088039867109636]\n",
      "Calculateing adj matrix using xy only...\n",
      "Run 5: radius [2.386485315859318, 3.3587570935487747], num_nbr [8.745847176079733, 20.088039867109636]\n",
      "Calculateing adj matrix using xy only...\n",
      "recommended radius =  2.8726212047040462 num_nbr=12.529900332225914\n",
      "radius= 2.8726212047040462 average number of neighbors for each spot is 12.529900332225914\n",
      " Cluster 1 has neighbors:\n",
      "Dmain  6 :  387\n",
      "Calculateing adj matrix using xy only...\n",
      "Calculateing adj matrix using xy only...\n",
      "Calculateing adj matrix using xy only...\n",
      "Run 1: radius [1.4142135381698608, 16.970561981201172], num_nbr [1.0, 245.5088339222615]\n",
      "Calculateing adj matrix using xy only...\n",
      "Run 2: radius [1.4142135381698608, 9.192387759685516], num_nbr [1.0, 89.87632508833923]\n",
      "Calculateing adj matrix using xy only...\n",
      "Run 3: radius [1.4142135381698608, 5.303300648927689], num_nbr [1.0, 32.936395759717314]\n",
      "Calculateing adj matrix using xy only...\n",
      "Run 4: radius [1.4142135381698608, 3.3587570935487747], num_nbr [1.0, 16.51590106007067]\n",
      "Calculateing adj matrix using xy only...\n",
      "Run 5: radius [2.386485315859318, 3.3587570935487747], num_nbr [7.6431095406360425, 16.51590106007067]\n",
      "Calculateing adj matrix using xy only...\n",
      "recommended radius =  2.8726212047040462 num_nbr=10.597173144876326\n",
      "radius= 2.8726212047040462 average number of neighbors for each spot is 10.597173144876326\n",
      " Cluster 2 has neighbors:\n",
      "Dmain  5 :  369\n",
      "Calculateing adj matrix using xy only...\n",
      "Calculateing adj matrix using xy only...\n",
      "Calculateing adj matrix using xy only...\n",
      "Run 1: radius [1.4142135381698608, 16.970561981201172], num_nbr [1.0, 352.93243243243245]\n",
      "Calculateing adj matrix using xy only...\n",
      "Run 2: radius [1.4142135381698608, 9.192387759685516], num_nbr [1.0, 124.62934362934362]\n",
      "Calculateing adj matrix using xy only...\n",
      "Run 3: radius [1.4142135381698608, 5.303300648927689], num_nbr [1.0, 42.513513513513516]\n",
      "Calculateing adj matrix using xy only...\n",
      "Run 4: radius [1.4142135381698608, 3.3587570935487747], num_nbr [1.0, 20.096525096525095]\n",
      "Calculateing adj matrix using xy only...\n",
      "Run 5: radius [2.386485315859318, 3.3587570935487747], num_nbr [8.714285714285714, 20.096525096525095]\n",
      "Calculateing adj matrix using xy only...\n",
      "recommended radius =  2.8726212047040462 num_nbr=12.53088803088803\n",
      "radius= 2.8726212047040462 average number of neighbors for each spot is 12.53088803088803\n",
      " Cluster 3 has neighbors:\n",
      "Dmain  5 :  662\n",
      "Dmain  4 :  660\n",
      "Calculateing adj matrix using xy only...\n",
      "Calculateing adj matrix using xy only...\n",
      "Calculateing adj matrix using xy only...\n",
      "Run 1: radius [1.4142135381698608, 16.970561981201172], num_nbr [1.0, 401.61581920903956]\n",
      "Calculateing adj matrix using xy only...\n",
      "Run 2: radius [1.4142135381698608, 9.192387759685516], num_nbr [1.0, 130.72693032015067]\n",
      "Calculateing adj matrix using xy only...\n",
      "Run 3: radius [1.4142135381698608, 5.303300648927689], num_nbr [1.0, 43.64783427495292]\n",
      "Calculateing adj matrix using xy only...\n",
      "Run 4: radius [1.4142135381698608, 3.3587570935487747], num_nbr [1.0, 20.51412429378531]\n",
      "Calculateing adj matrix using xy only...\n",
      "Run 5: radius [2.386485315859318, 3.3587570935487747], num_nbr [8.849340866290019, 20.51412429378531]\n",
      "Calculateing adj matrix using xy only...\n",
      "recommended radius =  2.8726212047040462 num_nbr=12.743879472693033\n",
      "radius= 2.8726212047040462 average number of neighbors for each spot is 12.743879472693033\n",
      " Cluster 4 has neighbors:\n",
      "Dmain  3 :  660\n",
      "Dmain  7 :  554\n",
      "Calculateing adj matrix using xy only...\n",
      "Calculateing adj matrix using xy only...\n",
      "Calculateing adj matrix using xy only...\n",
      "Run 1: radius [1.4142135381698608, 16.970561981201172], num_nbr [1.0, 290.8695652173913]\n",
      "Calculateing adj matrix using xy only...\n",
      "Run 2: radius [1.4142135381698608, 9.192387759685516], num_nbr [1.0, 109.32125603864735]\n",
      "Calculateing adj matrix using xy only...\n",
      "Run 3: radius [1.4142135381698608, 5.303300648927689], num_nbr [1.0, 40.20531400966183]\n",
      "Calculateing adj matrix using xy only...\n",
      "Run 4: radius [1.4142135381698608, 3.3587570935487747], num_nbr [1.0, 19.553140096618357]\n",
      "Calculateing adj matrix using xy only...\n",
      "Run 5: radius [2.386485315859318, 3.3587570935487747], num_nbr [8.6256038647343, 19.553140096618357]\n",
      "Calculateing adj matrix using xy only...\n",
      "recommended radius =  2.8726212047040462 num_nbr=12.27536231884058\n",
      "radius= 2.8726212047040462 average number of neighbors for each spot is 12.27536231884058\n",
      " Cluster 5 has neighbors:\n",
      "Dmain  3 :  662\n",
      "Dmain  2 :  369\n",
      "Calculateing adj matrix using xy only...\n",
      "Calculateing adj matrix using xy only...\n",
      "Calculateing adj matrix using xy only...\n",
      "Run 1: radius [1.4142135381698608, 16.970561981201172], num_nbr [1.0, 400.62158273381294]\n",
      "Calculateing adj matrix using xy only...\n",
      "Run 2: radius [1.4142135381698608, 9.192387759685516], num_nbr [1.0, 129.78705035971223]\n",
      "Calculateing adj matrix using xy only...\n",
      "Run 3: radius [1.4142135381698608, 5.303300648927689], num_nbr [1.0, 43.43597122302158]\n",
      "Calculateing adj matrix using xy only...\n",
      "Run 4: radius [1.4142135381698608, 3.3587570935487747], num_nbr [1.0, 20.450359712230217]\n",
      "Calculateing adj matrix using xy only...\n",
      "Run 5: radius [2.386485315859318, 3.3587570935487747], num_nbr [8.83884892086331, 20.450359712230217]\n",
      "Calculateing adj matrix using xy only...\n",
      "recommended radius =  2.8726212047040462 num_nbr=12.70935251798561\n",
      "radius= 2.8726212047040462 average number of neighbors for each spot is 12.70935251798561\n",
      " Cluster 6 has neighbors:\n",
      "Dmain  7 :  494\n",
      "Dmain  1 :  387\n",
      "Calculateing adj matrix using xy only...\n",
      "Calculateing adj matrix using xy only...\n",
      "Calculateing adj matrix using xy only...\n",
      "Run 1: radius [1.4142135381698608, 16.970561981201172], num_nbr [1.0, 411.8204697986577]\n",
      "Calculateing adj matrix using xy only...\n",
      "Run 2: radius [1.4142135381698608, 9.192387759685516], num_nbr [1.0, 133.06208053691276]\n",
      "Calculateing adj matrix using xy only...\n",
      "Run 3: radius [1.4142135381698608, 5.303300648927689], num_nbr [1.0, 44.33724832214765]\n",
      "Calculateing adj matrix using xy only...\n",
      "Run 4: radius [1.4142135381698608, 3.3587570935487747], num_nbr [1.0, 20.803691275167786]\n",
      "Calculateing adj matrix using xy only...\n",
      "Run 5: radius [2.386485315859318, 3.3587570935487747], num_nbr [8.953020134228188, 20.803691275167786]\n",
      "Calculateing adj matrix using xy only...\n",
      "recommended radius =  2.8726212047040462 num_nbr=12.906040268456376\n",
      "radius= 2.8726212047040462 average number of neighbors for each spot is 12.906040268456376\n",
      " Cluster 7 has neighbors:\n",
      "Dmain  4 :  554\n",
      "Dmain  6 :  494\n"
     ]
    }
   ],
   "source": [
    "adj_matrix = dict()\n",
    "for i in range(1, n_clusters + 1):\n",
    "    adj_matrix[i] = getXX(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SVGs for domain  1 : ['MBP', 'TMSB4X', 'TUBA1A', 'PLP1', 'FTH1', 'RPLP1', 'CNP', 'CRYAB', 'GFAP', 'RPL41', 'EEF1A1', 'FTL', 'RPL34', 'PTGDS', 'RPL28', 'RPL37', 'RPL37A', 'RPL32', 'RPS12', 'RPS27A', 'RPL13', 'S100B', 'RPL30', 'RPS15', 'RPL10', 'RPL19', 'TF', 'ITM2B', 'RPS28', 'RPS8', 'RPL13A', 'RPS21', 'RPS24', 'RPS23', 'APLP1', 'MOBP', 'RPLP2', 'RPL11', 'FAU', 'RPS27', 'TPT1', 'RPL21', 'RPS29', 'PTMA', 'RPS14', 'DPYSL2', 'RPL8', 'EIF1', 'RPL36', 'GPM6B', 'MTURN', 'RPS18', 'SCD', 'RPL29', 'RPL18', 'MARCKSL1', 'RPL35A', 'ATP5ME', 'CLDND1', 'RPL38', 'RPL4', 'SEPT4', 'MALAT1', 'RPS13', 'SERF2', 'C4orf48', 'APOD', 'RPL39', 'RPS15A', 'RPS19', 'GLUL', 'RPL35', 'PLEKHB1', 'RPS25', 'PPP1R14A', 'RPS4X', 'SPP1', 'QKI', 'RPL12', 'RNASE1', 'RPL22', 'GPRC5B', 'RPSA', 'RPL24', 'QDPR', 'RPL23A', 'MAG', 'RPS6', 'FEZ1', 'RPS3', 'CSRP1', 'B2M', 'PAQR6', 'NDRG2', 'RPS7', 'TUBB4A', 'RPS16', 'TSC22D4', 'RPS9', 'RPS5', 'SEPT7', 'PEA15', 'CD81', 'RPS10', 'MT2A', 'SELENOP', 'ERMN', 'SCD5', 'ABCA2', 'CLDN11', 'GSN', 'PPDPF', 'LINC00844', 'NENF', 'BCAS1', 'H3F3B', 'SIRT2', 'YWHAQ', 'H3F3A', 'EDIL3']\n",
      "Number of entries in the 'genes' column: 120\n",
      "SVGs for domain  2 : []\n",
      "Number of entries in the 'genes' column: 0\n",
      "SVGs for domain  3 : []\n",
      "Number of entries in the 'genes' column: 0\n",
      "SVGs for domain  4 : ['NEFL', 'NEFM', 'LGALS1', 'SCN1B']\n",
      "Number of entries in the 'genes' column: 4\n",
      "SVGs for domain  5 : []\n",
      "Number of entries in the 'genes' column: 0\n",
      "SVGs for domain  6 : ['NRGN', 'CCK', 'CHN1', 'SYT1', 'THY1', 'SLC17A7', 'ENC1', 'FBXL16', 'DKK3']\n",
      "Number of entries in the 'genes' column: 9\n",
      "SVGs for domain  7 : ['TMSB10', 'TUBB2A', 'PCP4']\n",
      "Number of entries in the 'genes' column: 3\n"
     ]
    }
   ],
   "source": [
    "#Set filtering criterials\n",
    "min_in_group_fraction=0.8\n",
    "min_in_out_group_ratio=1.0\n",
    "min_fold_change=1.2\n",
    "one_adj_filtered_info_ls = []\n",
    "for i in list(adj_matrix.keys()):\n",
    "    if (len(adj_matrix[i]) == 0):\n",
    "        continue\n",
    "    \n",
    "    target = i\n",
    "    \n",
    "    de_genes_info = rank_genes_groups(input_adata=raw,\n",
    "                                target_cluster=target,\n",
    "                                nbr_list=adj_matrix[target],\n",
    "                                label_col=\"mclust\",\n",
    "                                adj_nbr=True,\n",
    "                                log=True)\n",
    "    \n",
    "    \n",
    "    #Filter genes\n",
    "    de_genes_info=de_genes_info[(de_genes_info[\"pvals_adj\"]<0.05)]\n",
    "    \n",
    "    de_genes_info=de_genes_info[(de_genes_info[\"pvals_adj\"]<0.05) &\n",
    "                                (de_genes_info[\"in_out_group_ratio\"]>min_in_out_group_ratio) &\n",
    "                                (de_genes_info[\"in_group_fraction\"]>min_in_group_fraction) &\n",
    "                                (de_genes_info[\"fold_change\"]>min_fold_change) ]\n",
    "    de_genes_info=de_genes_info.sort_values(by=\"in_group_fraction\", ascending=False)\n",
    "    de_genes_info[\"target_dmain\"]=target\n",
    "    de_genes_info[\"neighbors\"]=str(adj_matrix[target])\n",
    "    print(\"SVGs for domain \", str(target),\":\", de_genes_info[\"genes\"].tolist())\n",
    "    genes_count = len(de_genes_info['genes'])\n",
    "    print(\"Number of entries in the 'genes' column:\", genes_count)\n",
    "    # print(de_genes_info)\n",
    "    one_adj_filtered_info_ls.append(de_genes_info)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "def adj_squre():\n",
    "    rst = dict()\n",
    "    for i in list(adj_matrix.keys()):\n",
    "        tmp = adj_matrix[i]\n",
    "        cur = set()\n",
    "        for j in tmp:\n",
    "            for k in adj_matrix[j]:\n",
    "                if i == k or k in tmp:\n",
    "                    continue\n",
    "                for m in adj_matrix[k]:\n",
    "                    if m == i or m in tmp:\n",
    "                        continue\n",
    "                    cur.add(m)\n",
    "        \n",
    "        rst[i] = cur\n",
    "    \n",
    "    for i in list(rst.keys()):\n",
    "        for j in list(rst[i]):\n",
    "            rst[j].add(i)\n",
    "    \n",
    "    for i in list(rst.keys()):     \n",
    "        rst[i] = list(rst[i])\n",
    "        \n",
    "    \n",
    "    print(rst)\n",
    "    print(type(rst))\n",
    "    return rst"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{1: [4], 2: [4], 3: [6], 4: [1, 2], 5: [7], 6: [3], 7: [5]}\n",
      "<class 'dict'>\n"
     ]
    }
   ],
   "source": [
    "#二阶邻域\n",
    "second_adj = adj_squre()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SVGs for domain  1 : ['MBP', 'PLP1', 'CNP', 'TUBA1A', 'CRYAB', 'GFAP', 'EEF1A1', 'FTL', 'PTGDS', 'RPL28', 'RPL37', 'RPS12', 'S100B', 'RPL30', 'RPL10', 'TF', 'RPS8', 'MOBP', 'APLP1', 'RPL11', 'RPS27', 'PTMA', 'GPM6B', 'MTURN', 'SCD', 'MARCKSL1', 'CLDND1', 'SEPT4', 'APOD', 'RPS15A', 'PLEKHB1', 'PPP1R14A', 'SPP1', 'QKI', 'RNASE1', 'GPRC5B', 'QDPR', 'MAG', 'FEZ1', 'CSRP1', 'B2M', 'SCGB2A2', 'PAQR6', 'TSC22D4', 'PEA15', 'MT2A', 'ERMN', 'SELENOP', 'ABCA2', 'CLDN11', 'GSN', 'PPDPF', 'NENF', 'LINC00844', 'BCAS1', 'H3F3B', 'SIRT2', 'H3F3A', 'EDIL3']\n",
      "Number of entries in the 'genes' column: 59\n",
      "SVGs for domain  2 : ['GFAP']\n",
      "Number of entries in the 'genes' column: 1\n",
      "SVGs for domain  3 : ['MT-ND3', 'GAPDH', 'RPL41', 'OLFM1', 'CKB', 'NDUFA4', 'COX6A1', 'NRGN', 'CALM1', 'CAMK2N1', 'TMSB4X', 'MT-ND5', 'ENC1', 'MTRNR2L12', 'VSNL1', 'YWHAH', 'CALM3', 'ATP1B1', 'PCSK1N', 'MT-ND4L', 'TPI1', 'UBC', 'ATP5MD', 'TSC22D1', 'ATP1A1', 'NEFL', 'GAP43', 'MTRNR2L8', 'MT-ATP8', 'SLC22A17', 'COX8A', 'NDUFA3', 'ATP1A3', 'PIN1', 'MEG3', 'SNCG', 'HOPX', 'MEF2C', 'NEFM']\n",
      "Number of entries in the 'genes' column: 39\n",
      "SVGs for domain  4 : ['MTRNR2L12', 'NRGN', 'MT-ND2', 'GAPDH', 'CLU', 'TMSB10', 'MT-ND3', 'MT-ND5', 'CKB', 'CALM2', 'CALM1', 'SNAP25', 'NDUFA4', 'COX6A1', 'TUBA1B', 'GNAS', 'RTN1', 'ATP1B1', 'TMSB4X', 'OLFM1', 'HSP90AA1', 'ACTB', 'UCHL1', 'MAP1B', 'COX4I1', 'NEFL', 'COX6C', 'STMN1', 'PPIA', 'YWHAH', 'VSNL1', 'HSP90AB1', 'RTN3', 'CCK', 'PEBP1', 'ACTG1', 'CALM3', 'HSPA8', 'SPARCL1', 'UBC', 'THY1', 'TPI1', 'COX5B', 'ATP5F1E', 'YWHAZ', 'TUBB2A', 'MT-ND4L', 'CAMK2N1', 'UQCRB', 'CFL1', 'PCSK1N', 'ENC1', 'SYT1', 'CHN1', 'PKM', 'YWHAG', 'GPM6A', 'NORAD', 'DYNLL1', 'SNRPN', 'SNCB', 'ATP5MD', 'BASP1', 'AP2M1', 'ATP5F1B', 'DNM1', 'EIF4A2', 'IDS', 'COX7C', 'GAP43', 'STMN2', 'RTN4', 'OAZ1', 'HINT1', 'UQCRQ', 'UQCR11', 'EEF1A2', 'VAMP2', 'LDHB', 'TSC22D1', 'ENO2', 'SLC17A7', 'PGAM1', 'NDRG4', 'ATP1A3', 'TSPAN7', 'ALDOC', 'BEX1', 'MDH1', 'ARPP19', 'GPX4', 'APP', 'COX8A', 'BEX3', 'STXBP1', 'PRNP', 'ITM2C', 'MTRNR2L8', 'LMO4', 'CLSTN1', 'ATP1A1', 'TMEM59L', 'SERPINI1', 'SYP', 'UQCRH', 'COX7A2', 'SOD1', 'PLD3', 'MORF4L1', 'ATP5MPL', 'C9orf16', 'PRKAR1B', 'GNG3', 'COX6B1', 'YWHAB', 'ATP5F1D', 'FBXL16', 'COPS9', 'SARAF', 'SLC22A17', 'CCNI', 'NDUFB2', 'KLC1', 'UQCR10', 'PRDX5', 'AES', 'NDFIP1', 'SPTBN1', 'MT-ATP8', 'SERINC1', 'NEFM', 'NDUFA3', 'PIN1', 'GDI1', 'STMN3', 'CABP1', 'CCDC85B', 'MAP2', 'GPI', 'NDUFS5', 'RAB3A', 'TSPYL1', 'PRDX2', 'FAIM2', 'RGS4', 'CLTC', 'ATP5MC3', 'MAP1A', 'MTCH1', 'GRIN1', 'PHYHIP', 'TUBB', 'NCS1', 'ATP5MF', 'SNCG', 'NSF', 'NCDN', 'RTL8C', 'PPP2R1A', 'PI4KA', 'ATP6V0B', 'DKK3', 'LGALS1', 'SPTAN1', 'GHITM', 'NAPB', 'MGST3', 'DSTN', 'ATP5PF', 'ATP5F1A', 'PREPL', 'ARF1', 'PPP3CA', 'CHCHD10', 'KIF5A', 'NDUFB1', 'TAGLN3', 'GNB1', 'PFN2', 'ATP2A2', 'PGK1', 'ATP2B2', 'TMEM160', 'COX7B', 'RAB6B', 'NPTN', 'ARF3', 'ARL6IP1', 'SCN1B', 'CHGA', 'SLC25A3', 'ATP5IF1', 'GABRA1', 'CTXN1', 'MEF2C', 'ATP6V1E1', 'SYN1', 'PINK1', 'B4GAT1', 'REEP5', 'SCG5', 'CTSB', 'TUBB4B', 'PTPRN', 'RHBDD2', 'GABARAPL1', 'SEZ6L2', 'RNF187', 'RUNDC3A', 'CALY', 'TUBA4A', 'LDHA', 'MOAP1', 'ATP5PO', 'DCLK1', 'TTC9B', 'GOT1', 'NCALD', 'CPLX1', 'DMTN', 'BEX2', 'NAPA', 'WSB2', 'CAMK2B', 'INA', 'ARHGDIG', 'APLP2']\n",
      "Number of entries in the 'genes' column: 227\n",
      "SVGs for domain  5 : ['CST3', 'ENC1', 'CALM1', 'MT-ND3', 'MT3', 'CLU', 'CAMK2N1', 'CCK', 'EEF1A1', 'MT-ND4L', 'PCSK1N', 'GPM6A', 'ARPP19', 'RTN4', 'RPL4', 'MALAT1', 'CTXN1', 'GRIN1', 'ITM2C', 'PHYHIP', 'MT-ATP8', 'APLP1', 'NCDN', 'C1orf61', 'PPP3CA', 'PRKCB', 'HOPX', 'NPTXR', 'HPCAL1', 'MEG3', 'PEA15', 'CAMK2A', 'MAST3', 'NGEF', 'C1QTNF4', 'ATP2B1', 'MT2A', 'GRIA2', 'PSD', 'CD63', 'FKBP1A', 'TTYH1', 'DDN', 'SERPINE2']\n",
      "Number of entries in the 'genes' column: 44\n",
      "SVGs for domain  6 : ['MBP', 'PLP1', 'GFAP', 'SCGB2A2', 'CRYAB']\n",
      "Number of entries in the 'genes' column: 5\n",
      "SVGs for domain  7 : ['TMSB10', 'SYT1', 'TUBA1B', 'SNAP25', 'MBP', 'STMN1', 'TUBB2A', 'MAP1B', 'NEFL', 'STMN2', 'PLP1', 'CCNI', 'NAPB', 'TTC9B', 'CPLX1', 'PCP4', 'SNCG', 'DCLK1', 'DIRAS2', 'LGALS1', 'SLC24A2', 'NEFM']\n",
      "Number of entries in the 'genes' column: 22\n"
     ]
    }
   ],
   "source": [
    "two_adj_filtered_info_ls = []\n",
    "for i in list(second_adj.keys()):\n",
    "    if (len(second_adj[i]) == 0):\n",
    "        continue\n",
    "    \n",
    "    target = i\n",
    "    \n",
    "    de_genes_info = rank_genes_groups(input_adata=raw,\n",
    "                                target_cluster=target,\n",
    "                                nbr_list=second_adj[target],\n",
    "                                label_col=\"mclust\",\n",
    "                                adj_nbr=True,\n",
    "                                log=True)\n",
    "    \n",
    "    \n",
    "    #Filter genes\n",
    "    de_genes_info=de_genes_info[(de_genes_info[\"pvals_adj\"]<0.05)]\n",
    "    filtered_info=de_genes_info\n",
    "    filtered_info=filtered_info[(filtered_info[\"pvals_adj\"]<0.05) &\n",
    "                                (filtered_info[\"in_out_group_ratio\"]>min_in_out_group_ratio) &\n",
    "                                (filtered_info[\"in_group_fraction\"]>min_in_group_fraction) &\n",
    "                                (filtered_info[\"fold_change\"]>min_fold_change)]\n",
    "    filtered_info=filtered_info.sort_values(by=\"in_group_fraction\", ascending=False)\n",
    "    filtered_info[\"target_dmain\"]=target\n",
    "    filtered_info[\"neighbors\"]=str(second_adj[target])\n",
    "    print(\"SVGs for domain \", str(target),\":\", filtered_info[\"genes\"].tolist())\n",
    "    genes_count = len(filtered_info['genes'])\n",
    "    print(\"Number of entries in the 'genes' column:\", genes_count)\n",
    "    # print(filtered_info)\n",
    "    two_adj_filtered_info_ls.append(filtered_info)\n",
    "    # filtered_info.to_csv(f'./my_result/filtered_info_{i}.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "def integration_one_two_adj_genes():\n",
    "    rst = np.array([])\n",
    "    for item in one_adj_filtered_info_ls:\n",
    "        rst = np.append(rst, item['genes'].values)\n",
    "    \n",
    "    for item in two_adj_filtered_info_ls:\n",
    "        rst = np.append(rst, item['genes'].values)\n",
    "    \n",
    "    return rst"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "integration_genes_array = integration_one_two_adj_genes()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "370"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import squidpy as sq\n",
    "from squidpy import gr, im, pl, read, datasets\n",
    "\n",
    "integration_genes_array = np.unique(integration_genes_array)\n",
    "len(integration_genes_array)\n",
    "pd.Index(integration_genes_array)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>I</th>\n",
       "      <th>pval_norm</th>\n",
       "      <th>var_norm</th>\n",
       "      <th>pval_norm_fdr_bh</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MBP</th>\n",
       "      <td>0.817429</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000092</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCGB2A2</th>\n",
       "      <td>0.749485</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000092</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-ND2</th>\n",
       "      <td>0.738200</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000092</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-ND3</th>\n",
       "      <td>0.713284</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000092</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PLP1</th>\n",
       "      <td>0.706320</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000092</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GFAP</th>\n",
       "      <td>0.645463</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000092</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MOBP</th>\n",
       "      <td>0.642673</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000092</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NRGN</th>\n",
       "      <td>0.629171</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000092</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FTH1</th>\n",
       "      <td>0.620985</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000092</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNAP25</th>\n",
       "      <td>0.612782</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000092</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                I  pval_norm  var_norm  pval_norm_fdr_bh\n",
       "MBP      0.817429        0.0  0.000092               0.0\n",
       "SCGB2A2  0.749485        0.0  0.000092               0.0\n",
       "MT-ND2   0.738200        0.0  0.000092               0.0\n",
       "MT-ND3   0.713284        0.0  0.000092               0.0\n",
       "PLP1     0.706320        0.0  0.000092               0.0\n",
       "GFAP     0.645463        0.0  0.000092               0.0\n",
       "MOBP     0.642673        0.0  0.000092               0.0\n",
       "NRGN     0.629171        0.0  0.000092               0.0\n",
       "FTH1     0.620985        0.0  0.000092               0.0\n",
       "SNAP25   0.612782        0.0  0.000092               0.0"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sq.gr.spatial_neighbors(raw, coord_type=\"generic\", delaunay=True)\n",
    "sq.gr.spatial_autocorr(raw, genes=pd.Index(integration_genes_array), use_raw=False, mode=\"moran\")\n",
    "raw.uns[\"moranI\"].head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "I列已成功保存为I_column.xlsx\n"
     ]
    }
   ],
   "source": [
    "import anndata as ad\n",
    "import pandas as pd\n",
    "# 从raw.uns中提取moranI\n",
    "moranI_data = raw.uns[\"moranI\"]\n",
    "\n",
    "# 将其转换为Pandas DataFrame（如果还不是DataFrame）\n",
    "if not isinstance(moranI_data, pd.DataFrame):\n",
    "    moranI_data = pd.DataFrame(moranI_data)\n",
    "\n",
    "# 提取I列\n",
    "I_column = moranI_data[[\"I\"]]  # 使用双中括号以保持DataFrame格式\n",
    "\n",
    "# 将I列存成Excel表格\n",
    "I_column.to_excel(\"../dataset/DLPFC/151673/GAADE_I_column.xlsx\", index=False)\n",
    "\n",
    "print(\"I列已成功保存为I_column.xlsx\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
